docs/LANGUAGE.md

```diff
@@ -0,0 +1,767 @@
+
+```
+rect
+# {"width": 2, "height": 4}
+```
+
+Many expressions are defined in terms of [Equality](#equality). These expressions can be thought of as assertions. The simplest type of assertion takes two scalar values:
+
+```opalog
+v :- 42 = "the meaning of life"
+```
+
+If we query for the contents of "v" we see the expression has been evaluated:
+
+```
+v
+# false
+```
+
+The order of operands in an equality expression does not matter:
+
+```opalog
+u :- "the meaning of life" = 42
+```
+
+The result is the same:
+
+```
+u
+# false
+```
+
+We can define rules in terms of [Variables](#variables) as well:
+
+```opalog
+t :- x = 42, y = 41, x > y
+```
+
+Multiple expressions are separated by the comma (",") character. In order for the rule to be true, all of the expressions in the rule must true for some set of variable bindings. There may be multiple sets of variable bindings that make the rule true. The body of a rule can be understood intuitively as "\<expression-1> AND \<expression-2> AND ... AND \<expression-N>".
+
+When we query for the contents of "t" we see the obvious result:
+
+```
+t
+# true
+```
+
+The order of expressions in a rule does not affect the document's content:
+
+```opalog
+s :- x > y, y = 41, x = 42
+```
+
+The query result is the same:
+
+```
+s
+# true
+```
+
+Opalog supports [References](#references) to nested documents. For example:
+
+```opalog
+sites :- [{"name": "prod"}, {"name": "smoke1"}, {"name": "dev"}]
+r :- sites[].name = "prod"
+```
+
+The rule "r" above asserts that there exists one document within the "sites"
+document which has the name "prod".
+
+The result:
+
+```
+r
+# true
+```
+
+We can generalize the example above with a rule that defines a set document instead of a boolean document:
+
+```opalog
+sites :- [{"name": "prod"}, {"name": "smoke1"}, {"name": "dev"}]
+q[] = name :- sites[].name = name
+```
+
+When we query for "q" we obtain a set of names:
+
+```
+q
+# ["prod", "smoke1", "dev"]
+```
+
+We can re-write the rule "r" from above to make use of "q". We will call the new rule "p":
+
+```opalog
+p :- "prod" = q[]
+```
+
+The result will be the same:
+
+```
+p
+# true
+```
+
+Rules which have arguments can be queried with input values:
+
+```
+q[] = "smoke2"
+# false
+
+q[] = "dev"
+# true
+```
+
+If you made it this far, congratulations. This section introduced the main
+aspsects of Opalog. The rest of this document provides more detail on various
+aspects of Opalog. Opalog's syntax is defined at the end of this document in
+the [Opalog Grammar](#grammar) section.
+
+## <a name="scalar-values"></a> Scalar Values
+
+Scalar values are the simplest type of term in Opalog. Scalar values can be
+strings, numbers, booleans, or null.
+
+Rules can be defined solely in terms of scalar values. This is useful for
+defining constants that are referenced in multiple places. For example:
+
+```opalog
+greeting   :- "Hello"
+max_height :- 42
+pi         :- 3.14159
+allowed    :- true
+sentinel   :- null
+```
```

Comment:
Just had a thought.
Instead of
pi :- 3.14159
we could write
pi = 3.14159 :- true
Then if we drop the body, which is unnecessary, we get..
pi = 3.14159

Then syntactically, every statement is a rule and a rule is of the form...
var[x] = y :- body
where body is optional, as is the [x]